hip/base/executor.hip.cpp

```diff
@@ -0,0 +1,198 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/executor.hpp>
+
+
+#include <iostream>
+
+
+#include <hip/hip_runtime.h>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+
+
+#include "hip/base/device_guard.hip.hpp"
+#include "hip/base/hipblas_bindings.hip.hpp"
+#include "hip/base/hipsparse_bindings.hip.hpp"
+
+
+namespace gko {
+
+
+std::shared_ptr<HipExecutor> HipExecutor::create(
+    int device_id, std::shared_ptr<Executor> master)
+{
+    return std::shared_ptr<HipExecutor>(
+        new HipExecutor(device_id, std::move(master)),
+        [device_id](HipExecutor *exec) {
+            delete exec;
+            if (!HipExecutor::get_num_execs(device_id)) {
+                device_guard g(device_id);
+                hipDeviceReset();
+            }
+        });
+}
+
+
+void OmpExecutor::raw_copy_to(const HipExecutor *dest, size_type num_bytes,
+                              const void *src_ptr, void *dest_ptr) const
+{
+    device_guard g(dest->get_device_id());
+    GKO_ASSERT_NO_HIP_ERRORS(
+        hipMemcpy(dest_ptr, src_ptr, num_bytes, hipMemcpyHostToDevice));
+}
+
+
+void HipExecutor::raw_free(void *ptr) const noexcept
+{
+    device_guard g(this->get_device_id());
+    auto error_code = hipFree(ptr);
+    if (error_code != hipSuccess) {
+#if GKO_VERBOSE_LEVEL >= 1
+        // Unfortunately, if memory free fails, there's not much we can do
+        std::cerr << "Unrecoverable HIP error on device " << this->device_id_
+                  << " in " << __func__ << ": " << hipGetErrorName(error_code)
+                  << ": " << hipGetErrorString(error_code) << std::endl
+                  << "Exiting program" << std::endl;
+#endif
+        std::exit(error_code);
+    }
+}
+
+
+void *HipExecutor::raw_alloc(size_type num_bytes) const
+{
+    void *dev_ptr = nullptr;
+    device_guard g(this->get_device_id());
+    auto error_code = hipMalloc(&dev_ptr, num_bytes);
+    if (error_code != hipErrorMemoryAllocation) {
+        GKO_ASSERT_NO_HIP_ERRORS(error_code);
+    }
+    GKO_ENSURE_ALLOCATED(dev_ptr, "hip", num_bytes);
+    return dev_ptr;
+}
+
+
+void HipExecutor::raw_copy_to(const OmpExecutor *, size_type num_bytes,
+                              const void *src_ptr, void *dest_ptr) const
+{
+    device_guard g(this->get_device_id());
+    GKO_ASSERT_NO_HIP_ERRORS(
+        hipMemcpy(dest_ptr, src_ptr, num_bytes, hipMemcpyDeviceToHost));
+}
+
+
+void HipExecutor::raw_copy_to(const CudaExecutor *src, size_type num_bytes,
+                              const void *src_ptr, void *dest_ptr) const
+{
+#if GINKGO_HIP_PLATFORM_NVCC == 1
+    device_guard g(this->get_device_id());
+    GKO_ASSERT_NO_HIP_ERRORS(hipMemcpyPeer(dest_ptr, this->device_id_, src_ptr,
+                                           src->get_device_id(), num_bytes));
+#else
+    GKO_NOT_SUPPORTED(HipExecutor);
```

Comment:
how about copying memory to the host memory and then to CUDA

Reply:
The problem is I don't really want to also add CUDA here, that feels like this could create quite a mess to have both CUDA and AMD symbols linked to the same file?

Reply:
I see.